README.md
# FileHeaderComment v2

Therefore that the author of the original [project](https://github.com/doi/fileheadercomment) did a perfect job but doesn't continue the development of this extension, this fork was created. It's main goal is a further development to add additional features to the extension.

This extension allow you to insert timestamp, copyright or any information to your file like comment below. The extension is able to detect multiple languages to select the correct comment prefix.

	/*
	 * Created on Tue Feb 18 2020
	 *
	 * Copyright (c) 2020 - Your Company
	 */

## Features

- insert defined parameter like `date`, `time`, `datetime`, `filename` (full list can be found [here](#variables))
- insert your own parameter and template
- detect files language to use correct comment prefix (list can be found [here](#language-list))
- define multiple templates

## Quick Jump
- [Updates](#updates)
- [Installation](#installation)
- [Setup](#setup)
- [Usage](#usage)
- [Variables](#variables)
- [Extension settings](#extension-settings)
- [Language list](#language-list)

## Updates
- [1.0.2] Updated install instructions

- [1.0.1] Fixed template variable issue

- [1.0.0] Added support for Makefile

## Installation
There are three ways to install this extension:
- VisualStudio Marketplace (or Extensiontab within VSC)
- manual install of vsix-file from releases
- git-clone and manual package export with vsce
```
$ git clone https://github.com/Zyzonix/VSC-FileHeaderComment-v2.git
$ npm install -g vsce
$ vsce package
```

## Setup
Firstly paste the default configuration to your `settings.js`: 

This file can be found under extension settings, keep in mind to edit the correct `settings.js` file in case you are using a SSH-connection or working in a WSL-window.

```
	"fileHeaderComment.parameter":{
		"*":{
			"commentbegin": "/*",
			"commentprefix": " *",
			"commentend": " */",
			"company": "Your Company"
		}
	},
	"fileHeaderComment.template":{
		"*":[
			"${commentbegin}",
			"${commentprefix} Created on ${date}",
			"${commentprefix}",
			"${commentprefix} Copyright (c) ${year} ${company}",
			"${commentend}"
		]
	}
```
## Usage
By default you don't have to set anything. It will detect most programming language for appropriate comment syntax.

Execute it from `Command Pallete` (menu View - Command Pallete... or ctrl+shift+p) then type command below:

1. `FileHeaderComment: Insert Default Template at Cursor`
2. `FileHeaderComment: Select from Available Templates`

The second command will show your available templates defined in Settings

If you want to set your own parameter and template (set from menu Preferences - User Settings), you can read explanation below

## Variables
You can use parameters below in your template

Parameter | Function | Example
---|---|---
`date` | current date | Thu May 26 2022
`dateeurop` | current date (formatted for europe) | 26.05.2022
`time` | current time | 10:23:52 PM
`time24h` | current time in 24 hour format | 22:23:52
`datetime`| current date + time | Thu May 26 2022 11:48:52 PM
`datetime24h` | current date + time in 24 hour format | Thu May 26 2022 23:48:52
`company` | "Your Company" | Company
`day` | day of the month | 26
`month` | current month | 05
`year` | current year | 2022
`hour` | current hour (24h) | 14
`minute` | current minute | 45
`second` | current second | 20
`filename` | filename | file.py
`filewithpath` | filename with filepath | path/to/file.py
`filewithpathandproject` | filename with filepath and project | project/path/to/file.py
`project` | projectname | project

## Language list
The following languages are automatically detected
- Swift
- Lua
- Perl
- Ruby
- Shellscript/Bash
- Yaml
- Makefile
- Python
- VB
- Clojure
- XML
- HTML
- JavaScript
- Java

If there's a language not on the list that you wish to use: set custom prefixes under parameters, see [here](#extension-settings) or request an update via opening an issue.


## Extension settings

Define all custom variables/paramenters in asterisk `*` like

```
"fileHeaderComment.parameter":{
	"*":{
		"company": "Your Company"
		"myvar1": "My Variable 1",
		"myvar2": "My Variable 2"
	}
}
```

Use your variable in template like (asterisk `*` will be default template)

```
"fileHeaderComment.template":{
	"*":[
		"${commentbegin}",
		"${commentprefix} Created on ${date}",
		"${commentprefix}",
		"${commentprefix} Copyright (c) ${year} ${company}",
		"${commentprefix} my variables are ${myvar1} and ${myvar2}",
		"${commentend}"
	]
}
```
You can define multiple templates, for instance template for MIT License

```
"fileHeaderComment.parameter":{
	"*":{
		"author": "Your Name"
	},
	"mit":{
		"license_mit":[
			"The MIT License (MIT)",
			" Copyright (c) ${year} ${author}",
			"",
			" Permission is hereby granted, free of charge, to any person obtaining a copy of this software",
			" and associated documentation files (the \"Software\"), to deal in the Software without restriction,",
			" including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,",
			" and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,",
			" subject to the following conditions:",
			"",
			" The above copyright notice and this permission notice shall be included in all copies or substantial",
			" portions of the Software.",
			"",
			" THE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED",
			" TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL",
			" THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,",
			" TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE."
		]
	}

},
"fileHeaderComment.template":{
	"*":[
		"${commentbegin}",
		"${commentprefix} Created by ${author}",
		"${commentend}"
	],
	"mit":[
		"${commentbegin}",
		"${commentprefix} Created on ${date}",
		"${commentprefix}",
		"${commentprefix} ${license_mit}",
		"${commentend}"
	]
}
```
You can use your `mit` template above by calling it through `Command Pallete` and choose `FileHeaderComment: Select from Available Templates`.

You can set also custom comment prefixes if your language is not automatically detected:

```
"fileHeaderComment.parameter":{
	"your_language":{
		"commentbegin": "#",
		"commentprefix": "#",
		"commentend": "#",
		"author": "Your name"
	}
},
"fileHeaderComment.template":{
	"your_language":{
		"${commentbegin}",
		"${commentprefix} Created by ${author}",
		"${commentend}"
	}
}
```

Readme version: 1.8


CHANGELOG.md
# Change Log

All notable changes to the "vsc-fileheadercomment-v2" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [1.0]

- Added support for Makefile

extension.js
/*
 * written by ZyzonixDev
 * published by ZyzonixDevelopments
 *
 * Copyright (c) 2022 ZyzonixDevelopments
 *
 * date created  | 30-05-2022 19:23:13
 * 
 * file          | extension.js
 * project       | vsc-fileheadercomment-v2
 * file version  | 1.0
 */


// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
var vscode = require('vscode');
var command = require('./command');

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {
	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "vsc-fileheadercomment-v2" is now active!');

	// The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json
    var disposable = vscode.commands.registerCommand('extension.insertFileHeaderComment', function () {
        // The code you place here will be executed every time your command is executed

        // Display a message box to the user
        // vscode.window.showInformationMessage('Hello World!');

        command.insertFileHeaderComment();
    });
    context.subscriptions.push(disposable);

    context.subscriptions.push(vscode.commands.registerCommand('extension.insertFileHeaderCommentOther', function(){
        command.insertFileHeaderCommentOther();
    }));
}
exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() {
}
exports.deactivate = deactivate;

command.js
/*
 * written by ZyzonixDev
 * published by ZyzonixDevelopments
 *
 * Copyright (c) 2022 ZyzonixDevelopments
 *
 * date created  | 27-05-2022 15:13:16
 * 
 * file          | command.js
 * project       | VSC-FileHeaderComment-v2
 * file version  | 1.0
 */

/*
 * Created on Sun Aug 07 2016
 *
 * The MIT License (MIT)
 * Copyright (c) 2016 Donna Iwan Setiawan
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/*
 * This project is a fork of https://github.com/doi/fileheadercomment
 *
 */

var vscode = require('vscode');
var path = require("path");

//import settings from settings.json
function insertFileHeaderComment(picked_template){
    console.log("Selected template: " + picked_template)
    var workspace = vscode.workspace,
        editor = vscode.window.activeTextEditor,
        // root = workspace.rootPath,
        prefix = 'fileHeaderComment',
        lang_id = editor.document.languageId,
        t_default = workspace.getConfiguration(prefix+".template").get('*'),
        // t_lang = workspace.getConfiguration(prefix+".template".get(lang_id),
        r_default = workspace.getConfiguration(prefix+".parameter").get('*'),
        r_lang = workspace.getConfiguration(prefix+".parameter").get(lang_id),
        template = [];
    if(picked_template){
        t_default = workspace.getConfiguration(prefix+".template").get(picked_template);
        r_default = workspace.getConfiguration(prefix+".parameter").get(picked_template);
        //if(tmp_r_default instanceof Object){
        //    Object.assign(r_default, tmp_r_default);
        //}
    }

    if(t_default instanceof Array){
        template = t_default;
    }else{
        template = [
            "${commentbegin}",
            "${commentprefix} Created on ${date}",
            "${commentprefix}",
            "${commentprefix} Copyright (c) ${year} ${company}",
            "${commentend}"
        ];
    }
    
    
    var workspace = vscode.workspace,
    
    // get workspace name
    rootPathSplit = workspace.rootPath.split("/"),
    projectName = rootPathSplit[rootPathSplit.length - 1],

    // get filename
    fileName = vscode.window.activeTextEditor.document.fileName.replace(/^.*[\\\/]/, '')

    // get full path of file (only project paths)
    fullFilePath = vscode.window.activeTextEditor.document.fileName,
    filePathSplit = fullFilePath.split(projectName),
    fileWithPath = filePathSplit[filePathSplit.length - 1].substring(1),
    fileWithPathAndProject = projectName + fileWithPath

    // format day / month to always two digits
    var date = new Date()
    if (date.getDate() < 10) {
        day = "0" + date.getDate().toString()
    } else {
        day = date.getDate()
    }
    if (date.getMonth()+1 < 10) {
        month = '0' + (date.getMonth()+1).toString()
    } else {
        month = date.getMonth()+1
    }

    // replace placeholders/paste data
    var h = (date.getHours()+"").padStart(2, '0'),
        m = (date.getMinutes()+"").padStart(2, '0'),
        s = (date.getSeconds()+"").padStart(2, '0'),
        
        replace = {            
            'date': date.toDateString(),                                // Weekday Month Day Year       | Thu May 26 2022
            'dateeurop': day + '-' + month + '-' + date.getFullYear(),  // Date                         | 26.05.2022
            'time': date.toLocaleTimeString(),                          // Time                         | 10:23:52 PM
            'time24h': h+':'+m+':'+s,                                   // Time in 24h format           | 22:23:52
            'day': day,                                                 // Day                          | 26
            'month': month,                                             // Month                        | 05
            'year': date.getFullYear(),                                 // Year                         | 2022
            'company': 'Your Company',                                  // Company / Team (specified as parameter)
            'filename': fileName,                                       // Filename                     | file.py                    
            'project': projectName,                                     // Project/Workspace name       | project
            'filewithpath': fileWithPath,                               // File with path               | /path/to/file.py
            'filewithpathandproject': fileWithPathAndProject,           // File with path and project   | project/path/to/file.py
            'hour': h,                                                  // Hour
            'minute': m,                                                // Minute
            'second': s,                                                // Second
        };
    
    replace = Object.assign(replace, {
        'commentbegin': '/*',
        'commentprefix': ' *',
        'commentend': ' */',
        'datetime': replace.date+ " "+replace.time,
        'datetime24h': replace.date+ " "+replace.time24h
    });
    replace.now = replace.datetime;
    replace.now24h = replace.datetime24h;
    Object.assign(replace, r_default);
    
    switch(lang_id){
        case "swift":
            replace.commentbegin = "/**";
            break;
        case "lua":
            Object.assign(replace, {
                'commentbegin': "--[[",
                'commentprefix': "--",
                'commentend': "--]]"
            });
            break;
        case "perl":
        case "ruby":
        case "shellscript":
        case "plaintext":
        case "yaml":
        case "makefile":
        case "python":
            Object.assign(replace, {
                'commentbegin': "#",
                'commentprefix': "#",
                'commentend': "#"
            });
            break;
        case "vb":
            Object.assign(replace, {
                'commentbegin': "'",
                'commentprefix': "'",
                'commentend': "'"
            });
            break;
        case 'clojure':
            Object.assign(replace, {
                'commentbegin': ";;",
                'commentprefix': ";",
                'commentend': ";;"
            });
            break;
        case "xml":
        case "html":
            Object.assign(replace, {
                'commentbegin': "<!--",
                'commentprefix': "",
                'commentend': "-->"
            });
            break;
    }
    Object.assign(replace, r_lang);

    if(!editor)
        return;
    replace = JSON.parse(JSON.stringify(replace));
    var s_template = template.join("\n"),
        escape = function(string) {
            return string.replace(/([.*+?^${}()|\[\]\/\\])/g, "\\$1");
        };
    for(var r in replace){
        var regexp = new RegExp(escape("${"+r+"}"), "gi"),
            replace_with = replace[r];
        if(replace_with.join){
            replace_with = replace_with.join("\n"+ replace.commentprefix);
        }else if(replace_with.replace){
            replace_with = replace_with.replace("\n", "\n"+ replace.commentprefix);
        }

        s_template = s_template.replace(regexp, replace_with);
    }

    //parse one more time
    //sometimes parameter has parameter inside it
    for(var r in replace){
        var regexp = new RegExp(escape("${"+r+"}"), "gi"),
            replace_with = replace[r];
        s_template = s_template.replace(regexp, replace_with);
    }

    //insert header comment at cursor
    editor.edit(function(edit){
        edit.insert(editor.selection.active, s_template+"\n");
    });
}
exports.insertFileHeaderComment = insertFileHeaderComment;
function insertFileHeaderCommentOther(){
    var workspace = vscode.workspace,
        prefix = 'fileHeaderComment',
        t_others = workspace.getConfiguration(prefix+".template");

    t_others = JSON.parse(JSON.stringify(t_others));
    var template_list = Object.keys(t_others);

    if(template_list.length == 0){
        template_list.push("*");
    }else if(template_list.indexOf("*") == -1){
        template_list.splice(0, 0, "*");
    }
    vscode.window.showQuickPick(template_list,{placeHolder: "Choose template"})
        .then(function(val){
            insertFileHeaderComment(val);
        });
}
exports.insertFileHeaderCommentOther = insertFileHeaderCommentOther;